watchtower/wtclient/client.go

```diff
@@ -0,0 +1,810 @@
+				c.processTask(task)
+			}
+		}
+	}
+}
+
+// processTask attempts to schedule the given backupTask on the active
+// sessionQueue. The task will either be accepted or rejected, afterwhich the
+// appropriate modifications to the client's state machine will be made. After
+// every invocation of processTask, the caller should ensure that the
+// sessionQueue hasn't been exhausted before proceeding to the next task. Tasks
+// that are rejected because the active sessionQueue is full will be cached as
+// the prevTask, and should be reprocessed after obtaining a new sessionQueue.
+func (c *TowerClient) processTask(task *backupTask) {
+	status, accepted := c.sessionQueue.AcceptTask(task)
+	if accepted {
+		c.taskAccepted(task, status)
+	} else {
+		c.taskRejected(task, status)
+	}
+}
+
+// taskAccepted processes the acceptance of a task by a sessionQueue depending
+// on the state the sessionQueue is in *after* the task is added. The client's
+// prevTask is always removed as a result of this call. The client's
+// sessionQueue will be removed if accepting the task left the sessionQueue in
+// an exhausted state.
+func (c *TowerClient) taskAccepted(task *backupTask, newStatus reserveStatus) {
+	log.Infof("Backup chanid=%s commit-height=%d accepted successfully",
+		task.id.ChanID, task.id.CommitHeight)
+
+	c.stats.taskAccepted()
+
+	// If this task was accepted, we discard anything held in the prevTask.
+	// Either it was nil before, or is the task which was just accepted.
+	c.prevTask = nil
+
+	switch newStatus {
+
+	// The sessionQueue still has capacity after accepting this task.
+	case reserveAvailable:
+
+	// The sessionQueue is full after accepting this task, so we will need
+	// to request a new one before proceeding.
+	case reserveExhausted:
+		c.stats.sessionExhausted()
+
+		log.Debugf("Session %s exhausted", c.sessionQueue.ID())
+
+		// This task left the session exhausted, set it to nil and
+		// proceed to the next loop so we can consume another
+		// pre-negotiated session or request another.
+		c.sessionQueue = nil
+	}
+}
+
+// taskRejected process the rejection of a task by a sessionQueue depending on
+// the state the was in *before* the task was rejected. The client's prevTask
+// will cache the task if the sessionQueue was exhausted before hand, and nil
+// the sessionQueue to find a new session. If the sessionQueue was not
+// exhausted, the client marks the task as ineligible, as this implies we
+// couldn't construct a valid justice transaction given the session's policy.
+func (c *TowerClient) taskRejected(task *backupTask, curStatus reserveStatus) {
+	switch curStatus {
+
+	// The sessionQueue has available capacity but the task was rejected,
+	// this indicates that the task was ineligible for backup.
+	case reserveAvailable:
+		c.stats.taskIneligible()
+
+		log.Infof("Backup chanid=%s commit-height=%d is ineligible",
+			task.id.ChanID, task.id.CommitHeight)
+
+		err := c.cfg.DB.MarkBackupIneligible(
```

Comment:
If this is a dust error, should we perhaps try with another tower? I see that the client only has access to one tower so far, so can add a TODO if so.

Reply:
atm the client only accepts a single policy, so if it produces dust it will also produce dust for any tower that also accepts that policy